### PR TITLE
Checks and ASTScanner module

### PR DESCRIPTION
**Pull request by loqo71la**

### CODE REVIEW

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv this is my progress for now

**Comment by JimmyRojas** [comment]
please add documentation for class, class member, methods, so on

**Comment by JimmyRojas** [comment]
(2) please add documentation for class, class member, methods, so on

**Comment by loqo71la** [comment]
@JimmyRojas @marcocdlv @ziur please review my changes.

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.sonar.check.Rule;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+import org.sonar.squidbridge.checks.SquidCheck;
+
+import org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey;
+
+/**
+ * Check the return type of the method is equal to the header method type
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java]: Checks

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        name = "Checking the return type of the method",
+        tags = Tags.CONVENTION
+)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
+@SqaleConstantRemediation("5min")
+@ActivatedByDefault
+public class CheckReturnTypeMethod extends SquidCheck<Grammar> {
+
+    /**
+     * Patron to show the message.
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java]: Stores a message template ...

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public static final String MESSAGE = "The type of the return value \"%s\", does not match the header \"%s\"";
+
+    /**
+     * Id code for the plugin.
+     */
+    public static final String CHECK_KEY = "S1003";
+
+    /**
+     * Is an axillary node, which help verify the next node.
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java]: Stores an AstNode instance which hepls ....

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+     * @param node where the checks are done.
+     */
+    @Override
+    public void visitNode(AstNode node) {
+        checkNode(node.getFirstDescendant(RuleKey.TYPE_METHOD));
+
+    }
+
+    /**
+     * Check if the new node is different from the first node. If it is true a
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java]: Checks ...

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+     * @param node used to compare.
+     */
+    private void checkNode(AstNode node) {
+        if (firstNode == null) {
+            firstNode = node;
+        } else if (!firstNode.getTokenValue().equals(node.getTokenValue())) {
+            getContext().createLineViolation(this,
+                    String.format(MESSAGE, node.getToken().getValue(), firstNode.getToken().getValue()), node);
+        }
+
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/CheckReturnTypeMethod.java]: remove this line

File: apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/Tags.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.fundacionjala.enforce.sonarqube.apex.checks;
+
+/**
+ * This class define tags constants.
```
Comment (JimmyRojas) [inline apex-checks/src/main/java/org/fundacionjala/enforce/sonarqube/apex/checks/Tags.java]: This class defines tags constants.

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexAstScanner.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+import org.fundacionjala.enforce.sonarqube.apex.parser.ApexParser;
+
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey.APEX_GRAMMAR;
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey.CLASS_NAME;
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey.CLASS_OR_INTERFACE_BODY_DECLARATION;
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey.METHOD_NAME;
+import static org.fundacionjala.enforce.sonarqube.apex.api.grammar.RuleKey.RESULT_TYPE;
+
+/**
+ * Utility class for scans a file and generate {@link SourceFile}
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/ApexAstScanner.java]: "for scans" ??

in order to scan

File: apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexMetric.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+package org.fundacionjala.enforce.sonarqube.apex.api;
+
+import org.sonar.squidbridge.measures.CalculatedMetricFormula;
+import org.sonar.squidbridge.measures.MetricDef;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+
+/**
+ * Define the type of metrics.
```
Comment (JimmyRojas) [inline apex-squid/src/main/java/org/fundacionjala/enforce/sonarqube/apex/api/ApexMetric.java]: This class defines ..

**Comment by JimmyRojas** [comment]
see some comments please..

**Comment by RockFercho** [comment]
@JimmyRojas @marcocdlv solved observations.

**Pull request merged**